refactor(app): add error handling for new stacker commands

sfoster1 opened the pull request:
Now that the commands are changing, some of the error recovery had to change too. Specifically `retrieve` doesn't have a labware ID param anymore, which is what retrieve error-recovery used to display a labware.

Instead, we can put that labware ID in the error info of the stacker error objects, and then we can get that in error recovery. That takes some typescript acrobatics to make sure that we can pull it out of the right kind of error, and that is greatly aided by adding the defined error types to the `RunTimeCommands`.

What I haven't done in this PR is add all the defined errors to all the runtime commands, but we really really should: this is the point of defined errors, that there's some errors where we know they can happen on certain commands, and we should write that down in the client that we write it down on the server.

## testing
- [x] do some retrieve error recovery passes and check that it functions
- [x] also check that command text works

sfoster1 committed:
refactor(api): put labware ids in stacker errors

Since we removed the labware ID from the params of some commands, we
need to replace them with something for error recovery. Let's do the
same thing we do for in-place liquid handling commands and stuff them in
an errorInfo dict.

sfoster1 committed:
refactor(shared-data): define more errors

Let's define errors a little better in the s-d ts types.

- We can parametrize RunTimeCommand by the defined errors that commands
should give
  - At some point we should really do this for all the commands, since
  the entire point of the defined error system is that there's a small
  number of them and we should write them all down explicitly
  - But for now I'm just going to do it for stacker
- We can have a nice type utility for getting a failed command
- We can create some flex stacker commands special that map the new
labware ID

sfoster1 committed:
refactor(app): handle new stacker errors

Nwo that the stacker retrieve command doesn't specify the labware ID we
have to get it a different way. We can do it from the errorInfo field on
the defined error, which is convenient. Unfortunately it's currently a
little inconvenient to pull that data out of the error types, so let's
refactor some of the utility code to make it a bit easier and a bit more
correct.

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/flex_stacker/common.py:
```diff
@@ -90,6 +91,12 @@
 )
 
 
+class FailedLabware(TypedDict, total=False):
```
nice

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/flex_stacker/common.py:
```diff
@@ -121,6 +132,8 @@ class FlexStackerHopperError(ErrorOccurrence):
     errorCode: str = ErrorCodes.STACKER_HOPPER_LABWARE_FAILED.value.code
     detail: str = ErrorCodes.STACKER_HOPPER_LABWARE_FAILED.value.detail
 
+    errorInfo: FailedLabware
```
`errorInfo: ErrorLocationInfo` in other errors (stall/overpressue). should we move it to `ErrorOccurrence` as optional with the union types? if not I think we should rename this?

TamarZanzouri commented on app/src/organisms/ErrorRecoveryFlows/hooks/useFailedLabwareUtils.ts:
```diff
@@ -185,12 +186,12 @@ export function useFailedLabwareUtils({
-  | Omit<PickUpTipRunTimeCommand, 'result'>
-  | Omit<MoveLabwareRunTimeCommand, 'result'>
-  | Omit<FlexStackerRetrieveRunTimeCommand, 'result'>
+  | Failed<AspirateRunTimeCommand>
```
nice

TamarZanzouri commented on app/src/organisms/ErrorRecoveryFlows/hooks/useFailedLabwareUtils.ts:
```diff
@@ -378,48 +379,63 @@ export function getFailedLabwareQuantity(
+    error != null &&
+    error.isDefined &&
+    [
+      'flexStackerStallOrCollision',
```
can we change this to use the const `STACKER_ERROR_KINDS` ?

sfoster1 replied:
We cannot, because these are not error kinds - these are the exception type keys. We'd have to make a new array or something, and I don't really want to because ts doesn't let you use `.includes()` to narrow unions anyway so we'd have to type them out

TamarZanzouri commented on app/src/organisms/ErrorRecoveryFlows/hooks/useRecoveryCommands.ts:
```diff
@@ -137,42 +142,41 @@ export function useRecoveryCommands({
+    return isInPlace(unvalidatedFailedCommand) &&
+      isTargetedError(unvalidatedFailedCommand.error) &&
+      // Paranoia: this value comes from the wire and may be unevenly implemented
+      typeof unvalidatedFailedCommand.error?.errorInfo?.retryLocation?.at(0) ===
```
thank you

TamarZanzouri reviewed (approved):
beautiful! a few comments if we can address but overall LGTM!

sfoster1 committed:
fix: proper false positive handling

When we skip retrieve in error recovery, we need to actually load the
"retrieved" labware into the module slot. To do that, we need to (1)
move the labware in retrieve's false-positive state update and (2)
actually use the false-positive state update when we resume

sfoster1 committed:
fix: incorrect multiple input

sfoster1 committed:
Merge branch 'edge' into exec-1394-add-ts-bindings
